File: examples/lookup.lua
#!/usr/bin/env lua
local function read_varint(f)
    local result, shift = 0, 0
    while true do
        local byte = f:read(1):byte()
        result = result | ((byte & 0x7F) << shift)
        if byte & 0x80 == 0 then return result end
        shift = shift + 7
    end
end

local function read_u16(f)
    local b = f:read(2)
    return b:byte(1) | (b:byte(2) << 8)
end

local function read_u32(f)
    local b = f:read(4)
    return b:byte(1) | (b:byte(2) << 8)
         | (b:byte(3) << 16) | (b:byte(4) << 24)
end

local function read_str(f)
    local len = f:read(1):byte()
    return f:read(len)
end

local function load(path)
    local f = assert(io.open(path or "blocklist.bin", "rb"))
    assert(f:read(4) == "IPBL")
    assert(f:read(1):byte() == 2)
    local timestamp = read_u32(f)

    local flag_count = f:read(1):byte()
    local flags = {}
    for i = 1, flag_count do flags[i] = read_str(f) end

    local cat_count = f:read(1):byte()
    local cats = {}
    for i = 1, cat_count do cats[i] = read_str(f) end

    local feed_count = read_u16(f)
    local feeds = {}

    for i = 1, feed_count do
        local name = read_str(f)
        local bs = f:read(1):byte()
        local co = f:read(1):byte()
        local fm = read_u32(f)
        local cm = f:read(1):byte()
        local rc = read_u32(f)

        local v4s, v4e = {}, {}
        local current = 0
        for _ = 1, rc do
            current = current + read_varint(f)
            local size = read_varint(f)
            local en = current + size
            if en <= 0xFFFFFFFF then
                v4s[#v4s + 1] = current
                v4e[#v4e + 1] = en
            end
        end

        local mf = {}
        for j = 1, #flags do
            if fm & (1 << (j - 1)) ~= 0 then mf[#mf+1] = flags[j] end
        end
        local mc = {}
        for j = 1, #cats do
            if cm & (1 << (j - 1)) ~= 0 then mc[#mc+1] = cats[j] end
        end

        feeds[i] = {
            name = name, bs = bs, co = co,
            flags = mf, cats = mc, v4s = v4s, v4e = v4e,
        }
    end
    f:close()
    return timestamp, feeds
end

local function bisect_right(arr, target)
    local lo, hi = 1, #arr + 1
    while lo < hi do
        local mid = math.floor((lo + hi) / 2)
        if arr[mid] <= target then lo = mid + 1 else hi = mid end
    end
    return lo
end

local function parse_ipv4(s)
    local a, b, c, d = s:match("^(%d+)%.(%d+)%.(%d+)%.(%d+)$")
    if not a then return nil end
    return (tonumber(a) << 24) | (tonumber(b) << 16)
         | (tonumber(c) << 8) | tonumber(d)
end

if #arg < 1 then
    io.stderr:write("Usage: lua lookup.lua <ip> [<ip> ...]\n")
    os.exit(1)
end

local _, feeds = load()

for _, ip in ipairs(arg) do
    local target = parse_ipv4(ip)
    if not target then
        print(ip .. ": invalid/unsupported IP")
        goto continue
    end

    local found = false
    for _, feed in ipairs(feeds) do
        local idx = bisect_right(feed.v4s, target) - 1
        if idx >= 1 and target <= feed.v4e[idx] then
            local score = (feed.bs / 200.0) * (feed.co / 200.0)
            local parts = { feed.name,
                string.format("score=%.2f", score) }
            if #feed.flags > 0 then
                parts[#parts+1] = "flags=" ..
                    table.concat(feed.flags, ",")
            end
            if #feed.cats > 0 then
                parts[#parts+1] = "cats=" ..
                    table.concat(feed.cats, ",")
            end
            print(ip .. ": " .. table.concat(parts, " | "))
            found = true
        end
    end

    if not found then print(ip .. ": no matches") end
    ::continue::
end

File: examples/lookup.java
import java.io.*;
import java.net.*;
import java.nio.*;
import java.nio.file.*;
import java.util.*;

public class lookup {
    record Feed(
        String name, int baseScore, int confidence,
        int flagsMask, int categoriesMask,
        long[] ipv4Starts, long[] ipv4Ends,
        long[] ipv6Starts, long[] ipv6Ends
    ) {}

    record Blocklist(
        long timestamp, String[] flags, String[] categories,
        Feed[] feeds
    ) {}

    static int pos;

    static long readVarint(byte[] d) {
        long result = 0;
        int shift = 0;
        while (true) {
            int b = d[pos++] & 0xFF;
            result |= (long) (b & 0x7F) << shift;
            if ((b & 0x80) == 0) return result;
            shift += 7;
        }
    }

    static String readStr(byte[] d) {
        int len = d[pos++] & 0xFF;
        String s = new String(d, pos, len);
        pos += len;
        return s;
    }

    static int readU16(byte[] d) {
        int v = (d[pos] & 0xFF) | ((d[pos + 1] & 0xFF) << 8);
        pos += 2;
        return v;
    }

    static long readU32(byte[] d) {
        long v = (d[pos] & 0xFFL) | ((d[pos+1] & 0xFFL) << 8)
               | ((d[pos+2] & 0xFFL) << 16)
               | ((d[pos+3] & 0xFFL) << 24);
        pos += 4;
        return v;
    }

    static Blocklist load(String path) throws IOException {
        byte[] d = Files.readAllBytes(Path.of(path));
        pos = 0;

        if (d[0] != 'I' || d[1] != 'P' || d[2] != 'B' || d[3] != 'L')
            throw new IOException("invalid magic");
        pos = 4;

        if (d[pos++] != 2) throw new IOException("bad version");
        long timestamp = readU32(d);

        int flagCount = d[pos++] & 0xFF;
        String[] flags = new String[flagCount];
        for (int i = 0; i < flagCount; i++) flags[i] = readStr(d);

        int catCount = d[pos++] & 0xFF;
        String[] cats = new String[catCount];
        for (int i = 0; i < catCount; i++) cats[i] = readStr(d);

        int feedCount = readU16(d);
        Feed[] feeds = new Feed[feedCount];

        for (int i = 0; i < feedCount; i++) {
            String name = readStr(d);
            int baseScore = d[pos++] & 0xFF;
            int confidence = d[pos++] & 0xFF;
            long fmask = readU32(d);
            int cmask = d[pos++] & 0xFF;
            long rangeCount = readU32(d);

            var v4s = new ArrayList<Long>();
            var v4e = new ArrayList<Long>();
            var v6s = new ArrayList<Long>();
            var v6e = new ArrayList<Long>();
            long current = 0;

            for (long r = 0; r < rangeCount; r++) {
                current += readVarint(d);
                long size = readVarint(d);
                long end = current + size;
                if (end <= 0xFFFFFFFFL) {
                    v4s.add(current);
                    v4e.add(end);
                } else {
                    v6s.add(current);
                    v6e.add(end);
                }
            }

            feeds[i] = new Feed(
                name, baseScore, confidence,
                (int) fmask, cmask,
                v4s.stream().mapToLong(Long::longValue).toArray(),
                v4e.stream().mapToLong(Long::longValue).toArray(),
                v6s.stream().mapToLong(Long::longValue).toArray(),
                v6e.stream().mapToLong(Long::longValue).toArray()
            );
        }

        return new Blocklist(timestamp, flags, cats, feeds);
    }

    static int bisectRight(long[] arr, long target) {
        int lo = 0, hi = arr.length;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (arr[mid] <= target) lo = mid + 1;
            else hi = mid;
        }
        return lo;
    }

    static String formatMatch(Blocklist bl, Feed f) {
        double score = (f.baseScore / 200.0)
                     * (f.confidence / 200.0);
        var parts = new ArrayList<>(
            List.of(f.name, String.format("score=%.2f", score))
        );

        var matched = new StringJoiner(",");
        for (int i = 0; i < bl.flags.length; i++)
            if ((f.flagsMask & (1 << i)) != 0)
                matched.add(bl.flags[i]);
        if (matched.length() > 0)
            parts.add("flags=" + matched);

        var cats = new StringJoiner(",");
        for (int i = 0; i < bl.categories.length; i++)
            if ((f.categoriesMask & (1 << i)) != 0)
                cats.add(bl.categories[i]);
        if (cats.length() > 0)
            parts.add("cats=" + cats);

        return String.join(" | ", parts);
    }

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.err.println("Usage: java lookup <ip> [<ip> ...]");
            System.exit(1);
        }

        Blocklist bl = load("blocklist.bin");

        for (String arg : args) {
            InetAddress addr;
            try {
                addr = InetAddress.getByName(arg);
            } catch (Exception e) {
                System.out.println(arg + ": invalid IP");
                continue;
            }

            boolean found = false;
            if (addr instanceof Inet4Address) {
                byte[] bytes = addr.getAddress();
                long target = ((bytes[0] & 0xFFL) << 24)
                    | ((bytes[1] & 0xFFL) << 16)
                    | ((bytes[2] & 0xFFL) << 8)
                    | (bytes[3] & 0xFFL);
                for (Feed f : bl.feeds) {
                    int idx = bisectRight(f.ipv4Starts, target) - 1;
                    if (idx >= 0 && target <= f.ipv4Ends[idx]) {
                        System.out.printf(
                            "%s: %s%n", arg, formatMatch(bl, f)
                        );
                        found = true;
                    }
                }
            }

            if (!found) System.out.println(arg + ": no matches");
        }
    }
}
